### PR TITLE
Better deletion of metadata in s3queue

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Delete table metadata in keeper in Storage S3(Azure)Queue only after last table using this metadata is dropped.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [x]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [x]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 03f44e8920b39d9546a6750bebdd8c636c35986d with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/72810/03f44e8920b39d9546a6750bebdd8c636c35986d/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/b364af07f9253814bd808cae88822aed6a3a2d0b/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Storages/ObjectStorageQueue/ObjectStorageQueueMetadataFactory.cpp [hunk 1]
```diff
@@ -27,34 +28,36 @@ ObjectStorageQueueMetadataFactory::FilesMetadataPtr ObjectStorageQueueMetadataFa
...
 {
     std::lock_guard lock(mutex);
     auto it = metadata_by_path.find(zookeeper_path);
 
     if (it == metadata_by_path.end())
         throw Exception(ErrorCodes::BAD_ARGUMENTS, "Metadata with zookeeper path {} does not exist", zookeeper_path);
 
-    chassert(it->second.ref_count > 0);
-    if (--it->second.ref_count == 0)
+    const size_t registry_size = it->second->unregister(storage_id);
```
Comment (SmitaRKulkarni) [inline src/Storages/ObjectStorageQueue/ObjectStorageQueueMetadataFactory.cpp:48]: maybe we should move `registerIfNot` and `unregister` inside the `try catch` block below, as they can fail for keeper fails. what do you think?

File: src/Storages/ObjectStorageQueue/ObjectStorageQueueMetadata.cpp [hunk 2]
```diff
@@ -438,6 +439,164 @@ ObjectStorageQueueTableMetadata ObjectStorageQueueMetadata::syncWithKeeper(
...
+                {
+                    LOG_TRACE(log, "Table {} is already registered", self.table_id);
+                    return;
+                }
+            }
+        }
+
+        LOG_TRACE(log, "Adding {} to registry", self.table_id);
+
+        if (node_exists)
```
Comment (SmitaRKulkarni) [inline src/Storages/ObjectStorageQueue/ObjectStorageQueueMetadata.cpp:519]: Can be moved to the above if block

File: tests/integration/test_storage_s3_queue/test.py [hunk 3]
```diff
@@ -2301,3 +2301,127 @@ def check_string_settings(node, settings):
...
+    dst_table_name = f"{table_name}_dst"
+    keeper_path = f"/clickhouse/test_{table_name}"
+    files_path = f"{table_name}_data"
+    files_to_generate = 1000
+
+    node1.query(f"DROP DATABASE IF EXISTS {db_name}")
+    node2.query(f"DROP DATABASE IF EXISTS {db_name}")
+
+    node1.query(
+        f"CREATE DATABASE {db_name} ENGINE=Replicated('/clickhouse/databases/replicateddb', 'shard1', 'node1')"
```
Comment (SmitaRKulkarni) [inline tests/integration/test_storage_s3_queue/test.py:2321]: we already use this for some other test, the integration test fail looks related to this.

**Review by SmitaRKulkarni — COMMENTED** [review]
Rest all LGTM

File: src/Storages/ObjectStorageQueue/ObjectStorageQueueMetadataFactory.cpp [hunk 4]
```diff
@@ -30,42 +31,67 @@ ObjectStorageQueueMetadataFactory::FilesMetadataPtr ObjectStorageQueueMetadataFa
         auto & metadata_from_keeper = it->second.metadata->getTableMetadata();
 
         metadata_from_table.checkEquals(metadata_from_keeper);
-
-        it->second.ref_count += 1;
     }
+
+    it->second.metadata->registerIfNot(storage_id);
```
Comment (SmitaRKulkarni) [inline src/Storages/ObjectStorageQueue/ObjectStorageQueueMetadataFactory.cpp:36]: no need for try catch block here ?

**Review by SmitaRKulkarni — APPROVED** [review]

File: src/Storages/ObjectStorageQueue/ObjectStorageQueueMetadataFactory.cpp (hunk 4 above)
Comment (kssenii) [reply src/Storages/ObjectStorageQueue/ObjectStorageQueueMetadataFactory.cpp:36]: No, it is ok to fail table creation in this case, we anyway to other stuff in keeper in constructor, not only this, without them we cannot start.

**Comment by kssenii** [comment]
> Integration tests (asan, old analyzer) [6/6] — fail: 1

test_storage_iceberg/test.py::test_evolved_schema_simple[True-hdfs-2]

**Pull request merged**